mds/apis/agency_api/v0_x/compliance.py

```diff
@@ -0,0 +1,109 @@
+from rest_framework import serializers
+from rest_framework import viewsets
+
+from django.db.models import Prefetch
+
+from mds import models
+
+from datetime import datetime
+import pytz
+
+
+class ComplianceSerializer(serializers.ModelSerializer):
+    rules_id = serializers.SerializerMethodField()
+    compliances = serializers.SerializerMethodField()
+
+    class Meta:
+        model = models.Policy
+        fields = ("id", "rules_id", "compliances")
+
+    def get_rules_id(self, policy):
+        return [rule["rule_id"] for rule in policy.rules]
+
+    def get_compliances(self, policy):
+        final_compliance = []
+        for compliance in policy.compliances_pref:
+            for current_compliance in final_compliance:
+                if current_compliance["rule_id"] == compliance.rule:
+                    for match in current_compliance["matches"]:
+                        if match["geography"] == str(compliance.geography):
+                            match["measured"] += 1
+                            break
+                    else:
+                        current_compliance["matches"].append(
+                            {"geography": str(compliance.geography), "measured": 1}
+                        )
+                    current_compliance["vehicles_in_violation"].append(
+                        str(compliance.vehicle.id)
+                    )
+                    current_compliance["total_violations"] += 1
+                    break
+            else:
+                final_compliance.append(
+                    {
+                        "rule_id": compliance.rule,
+                        "matches": [
+                            {"geography": str(compliance.geography), "measured": 1}
+                        ],
+                        "vehicles_in_violation": [str(compliance.vehicle.id)],
+                        "total_violations": 1,
+                    }
+                )
+        return final_compliance
+
+
+class ComplianceViewSet(viewsets.ModelViewSet):
+    queryset = models.Policy.objects
+    serializer_class = ComplianceSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        provider_id = self.request.GET.get("provider_id")
+        end_date = self.request.GET.get("end_date")
+        filters = {}
+        if provider_id and end_date:
+            filters["compliances__vehicle__provider__id"]: provider_id
+            filters["compliances__start_date__lte"]: end_date
```

Comment:
didn't you forget the compliances with end_date=None ?

Reply:
Thanks to the .exclude it delete all the end_date less that the end_date put in the call API. So he keep all the end_date higher and None. I make a little bit time to understand that yesterday but it's tested and approved :)

Reply:
If you want i can add that in the test_compliance

Reply:
i put that in test_compliance ;)